flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java

```diff
@@ -185,16 +186,21 @@ public static CheckpointStorage load(
                     CheckpointingOptions.SAVEPOINT_DIRECTORY, defaultSavepointDirectory.toString());
         }
 
-        // Legacy state backends always take precedence
-        // for backwards compatibility.
-        if (configuredStateBackend instanceof CheckpointStorage) {
+        // Legacy state backends always take precedence for backwards compatibility.
+        StateBackend rootStateBackend =
+                (configuredStateBackend instanceof DelegatingStateBackend)
+                        ? ((DelegatingStateBackend) configuredStateBackend)
+                                .getDelegatedStateBackend()
+                        : configuredStateBackend;
+
+        if (rootStateBackend instanceof CheckpointStorage) {
```

Comment:
nit: wouldn't it be better to replace this with a default method in `StateBackend`:

```
interface StateBackend {
...
default boolean isCheckpointStorage() {
  return this instanceof CheckpointStorage;
}
...
}
```
and overload this method in `DelegatingStateBackend`/`ChangelogStateBackend`?

Reply:
Maybe we can separate this to a different PR? I think this should be a follow-up of 
"FLINK-19463 Disentangle StateBackends from Checkpointing"

and find all places of `if (rootStateBackend instanceof CheckpointStorage)`?